### ISSUE TITLE
Docstring sweep: preprocessing and the pipeline

### ISSUE THREAD

**Issue by dcelisgarza** [dcelisgarza/PortfolioOptimisers.jl#366]
## Question

**Sweep the 27 type boxes below against the three checks the map defines.**

The rules live on the map, [Systematic docstring sweep](https://github.com/dcelisgarza/PortfolioOptimisers.jl/issues/351). Read *The rule for one box* there before the first box. The map holds every rule; this ticket holds only the boxes and the files that carry them.

27 boxes over 3 files, 3986 source lines. A box is one type. A method docstring is swept in the box of the family that it serves, so sweep the **file**, not only the type.

### `src/03_Preprocessing.jl` — 2314 lines

- [x] `AbstractAssetSelector` — `none`
- [x] `AbstractPreprocessingEstimator` — `none`
- [x] `AbstractPreprocessingResult` — `none`
- [x] `AbstractPricesPreprocessingEstimator` — `none`
- [x] `AbstractPricesPreprocessingResult` — `none`
- [x] `AbstractPricesResult` — `none`
- [x] `AbstractReturnsPreprocessingEstimator` — `none`
- [x] `AbstractReturnsPreprocessingResult` — `none`
- [x] `AbstractReturnsResult` — `none`
- [x] `AssetSelectorResult` — `none`
- [x] `Imputer` — `none`
- [x] `ImputerResult` — `none`
- [x] `MissingDataFilter` — `none`
- [x] `MissingDataFilterResult` — `none`
- [x] `PricesResult` — `none`
- [x] `PricesToReturns` — `none`
- [x] `ReturnsResult` — `none`
- [x] `TrainTestSplit` — `none`
- [x] `TrainTestSplitResult` — `none`

### `src/23_Pipeline/01_Base_Pipeline.jl` — 652 lines

- [x] `AbstractPipelineEstimator` — `none`
- [x] `AbstractPipelineResult` — `none`
- [x] `PipelineContext` — `none`
- [x] `PipelineStep` — `none`
- [x] `PipelineUncertaintySets` — `none`
- [x] `TargetedConstraint` — `none`

### `src/23_Pipeline/03_Pipeline.jl` — 1020 lines

- [x] `Pipeline` — `none`
- [x] `PipelineResult` — `none`

## Done when

Every box above carries a tick and a reference key, or a tick and `none`. A box whose source should exist but cannot be found stays **unticked**, records `unknown`, and adds one row to the map's *Citations not found* section; the ticket then stays open until the maintainer names the work.

A defect found in a box is fixed in the same session and is never ticketed. Post the resolution as a comment, close the ticket, and add one line to the map's *Decisions so far*.

**Comment by dcelisgarza** [comment]
## Resolution

**27 boxes, all ticked `none`, and check 2 found seven defects — one of them in the code, not in the prose.**

### Check 1 — references: `none` for all 27, checked against the book, not guessed

Neither file names a work in its source comments; the only URLs are package links (`Impute.jl`, `TimeSeries.jl`). Two candidate sources were examined and both were ruled out **by reading**:

- **The pipeline family is not scikit-learn's.** ADR 0028 rejects it in as many words: "strict sklearn-style linear piping was rejected because the fan-in at optimisation would force a mega-step." The shape that shipped is a linear step list over an accumulating blackboard, which follows no published work.
- **The holdout is in no chapter of `cajas2025`.** Chapter 15 (*Backtesting Process*) enumerates the three protocols the book uses — 15.1 walk-forward, 15.2 k-fold cross-validation, 15.3 combinatorial purged cross-validation — and the single train/test holdout is not among them. For other splitters the book's own footnote points at scikit-learn's documentation rather than at a work.

One term of art in the file **does** have a source, and it belongs to another box. `safe_index` calls the rows between the two windows **embargoed**. Footnote 3 of section 15.3 attributes the embargo to **chapter 7 of López de Prado (2018)**, and the CPCV method itself to chapter 12 — which is `CombinatorialCrossValidation`'s source, in [the cross-validation ticket](https://github.com/dcelisgarza/PortfolioOptimisers.jl/issues/383), not this one. `TrainTestSplit`'s gap is declared by its two sizes and derives nothing from a label horizon, so it borrows the word without implementing the rule. `safe_index` now says so and links [`CombinatorialCrossValidation`](@ref).

### Check 2 — formula against code, with numbers: seven defects

**1. `prices_to_returns` does not compute excess returns.** Its `# Mathematical definition` claimed `r̃ = r − b` when a benchmark is given, and a `# Details` bullet repeated it: "If `B` is not `nothing`, it is subtracted from asset returns." Both are false. On a 4×2 price series with a single-column benchmark rising 2 % a period:

- `rr.X = [0.10 0.05]` — the **raw** asset returns, equal to `values(percentchange(P, :simple))` exactly.
- `rr.B = 0.02` — the benchmark, converted by the same rule and **carried**, not spent.
- `X == raw` is `true`; `X == raw .- b` is `false`.

The subtraction lives in `returns_result_picker(rd, brt)` and runs only for a benchmark-tracking optimisation. Subtracting here would double-subtract there. The docstring now says which function subtracts and when.

**2. `ReturnsResult`'s `ivpa` validation is wrong on both the guard and the bound.** The docstring said "If `ivpa` is not `nothing`, `all(x -> x >= 0, ivpa)`". Neither half holds:

- `ReturnsResult(; nx = ["A"], X = [0.1; 0.2;;], ivpa = -3.0)` **constructs**, holding `ivpa = -3.0`. The check sits inside the `!isnothing(iv)` branch, so an `ivpa` with no `iv` is never validated.
- With an `iv` present, `ivpa = 0.0` **throws `DomainError`**. The predicate is `assert_nonempty_gt0_finite_val` — strictly `> 0`, not `>= 0`. `PricesResult`'s own bullet already said `> 0`; the two carriers disagreed on paper only.

`prices_to_returns` carried a verbatim copy of the same wrong bullet. Both now state the guard and the strict bound.

**3. A single-column benchmark made an asset view throw — a code defect, now fixed.** `port_opt_view(::PricesResult, timestamps, assets)` chose how to slice `B` by testing `length(j) >= 1`, which asks about the **asset index** and not about the benchmark's width. A `PricesResult` whose `B` has one shared column, viewed on more than one asset, indexed `colnames(B)[j]` past its single column:

```
port_opt_view(pr, ts[2:3], [1, 3])   # B one column, X three assets
before: BoundsError
after:  B keeps its column :BM, rows 2:3 = [51.0, 52.0]
```

A per-asset benchmark is unaffected and still slices: columns `[:A, :C]`, values `[4.0 6.0; 7.0 9.0]`. The test is now `B`'s own width against `size(values(X), 2)`, which is the only shape distinction the constructor admits. The sibling arity carried the same guard with **two identical branches**; it is now the one-line form it always was.

This is a defect fix, not an API change: the docstring already documents `B` as `(observations × 1)` or `(observations × assets)`, and one of the two documented shapes could not survive an asset view.

**4. The same view's `# Details` described the wrong arity.** It said "`ivpa` is per-asset and passes through unchanged" and "a static feature matrix `Z` … passes through unchanged". True of the observation-only arity, false of the asset one: on a 3-asset carrier viewed at `[1, 3]`, `ivpa` goes `[1.0, 2.0, 3.0] → [1.0, 3.0]` and a static `Z` goes 3 rows → 2. The asset index `j` was undocumented altogether. Both arities are now described, including that `F` keeps every column because factors are a separate axis.

**5. `first_duplicate` returns an element, not a set.** Its `# Returns` read "`seen::Set{String}`: A collection of names." It returns the first repeated element (`"a"` for `["a","b","a","c"]`) or `nothing` when every element is unique — which is what the caller interpolates into the uniqueness error.

**6. `predict(::PipelineResult, …)` had been mangled by a mechanical rename.** Every occurrence of "window" had become "test_idx", leaving "an unseen data test_idx", "the test test_idx", and "Observation test_idx into the rows". Restored, and while there: `cols` was undocumented, and the `VecVecInt` arity returns one result per window, which nothing said.

**7. `returns_result_picker`'s field list was incomplete and silent on the benchmark.** It claimed to preserve `nx, nf, F, ts, iv, ivpa` — omitting `nz`/`Z`, which it does preserve, and not mentioning that `nb` and `B` come back `nothing`. That is what makes the adjustment idempotent: `returns_result_picker(returns_result_picker(rd, true), true) === ` its argument, verified.

### Check 3 — the standard

- `PipelineStep`'s `reads` is a tuple of slots; its field docstring said "Slot".
- `TargetedConstraint`'s `# Constructors` block was missing "Keywords correspond to the struct's fields."
- No `# Constructors` section was added to `AssetSelectorResult`, `TrainTestSplitResult`, `MissingDataFilterResult`, `ImputerResult` or `PipelineResult`: none of them has a keyword constructor (`AssetSelectorResult(; nx = ["A"])` is a `MethodError`), so they are built by `fit_preprocessing` alone.
- Every summary sentence already stands alone; none was rewritten, so no catalogue entry moves.

### Gates

Run in a cold-loaded session, `BLAS.set_num_threads(1)`:

| File | Pass | Fail |
| --- | --- | --- |
| `test_06_preprocessing.jl` | 81 | 0 |
| `test_37_train_test_split.jl` | 66 | 0 |
| `test_30_pipeline_base.jl` | 87 | 0 |
| `test_31_pipeline_preprocessing.jl` | 50 | 0 |
| `test_32_pipeline_fit.jl` | 94 | 0 |
| `test_38_pipeline_routing.jl` | 180 | 0 |
| `test_26_docs.jl` — References | 450 | 0 |

`Base.undocumented_names` is 0 public and 0 private. `JuliaFormatter.format_file` reports all three files already formatted.

**Two gates were not run.** The full `test_26_docs.jl` wedged at the shared kaimon gate at the 60-minute timeout under contention from the concurrent sweep sessions; its *References* and *Docs completeness* testsets were run directly instead, and they are the two this ticket can affect — the diff adds no `# References` bullet, no `ref_dict` entry and no bibliography block, so no API page changes citing state. `test_33`/`test_34` were not run: the only change in their area is docstring prose.

**Not committed.** Two sibling sweep sessions hold uncommitted work in the same checkout (`src/01_Base.jl`, `src/02_Tools.jl`, `src/04_PosdefMatrix.jl`–`src/07_MatrixProcessing.jl`, `docs/src/References.bib`, two API pages). The pre-commit hook stashes unstaged files, so committing here can revert a peer's edit mid-flight. The three files of this ticket are disjoint from theirs and commit cleanly on their own.
